Streams/MemoryStreamReader.cpp

```diff
@@ -71,5 +71,5 @@ std::unique_ptr<SeekableStreamReader> MemoryStreamReader::Slice(uint64_t sliceSt
 		throw std::runtime_error("Unable to create a slice of memory stream. Requested slice is outside bounds of underlying stream.");
 	}
 
-	return std::make_unique<MemoryStreamReader>((void*)&streamBuffer[sliceStartPosition], static_cast<std::size_t>(sliceLength));
+	return MemoryStreamReader((void*)(&streamBuffer[sliceStartPosition]), static_cast<std::size_t>(sliceLength));
```

Comment:
Ahh, `static_cast` can't cast away `const`. The `streamBuffer` points to a `const` buffer.

I believe the correct solution here, is to add the necessary `const` to the constructors, as this data is expected to be `const`. You may need to add a few as the `const`-ness spreads.

Reply:
I just looked into const_cast as I didn't know it existed. We need to figure out how to change a const void const * buffer to a const char const * buffer, which I haven't figured out yet. 

I think switching the constructor to accept a const char const * buffer instead would fix the issue. Is there any reason to make the MemoryStreamReader's constructor a void* if the underlying buffer is going to be a char*?

-Brett

Reply:
There's very little reason to use `const_cast`. I think it's included more because the operation was allowed by the old C-style cast. I suppose it might be necessary in rare circumstances, where `const` was not used properly in library code you need to interface with.

In terms of the conversion, assuming the const-ness doesn't change, you can just use `static_cast`:
```
static_cast<const char*>(someConstVoidPtr)
```

The constructor should accept a `const void*`, rather than a `void*`. That's the real problem. As long as the const-ness matches, you can `static_cast` it.

I think using `const void*` in the interface makes sense, since it allows easy cast free conversion from a pointer to any type. The memory could come from `new`, or from `malloc`, or allocated with a `vector`, over overlayed on top of a `struct`. The `MemoryStreamReader` shouldn't know and shouldn't care. It's just memory that needs to be used as an input buffer so it can be sliced and diced into something new.

Reply:
Okay, thanks. It actually never occurred to me that it would be legal to place a const inside a static_cast like 'static_cast<const char* const>'. Duh.

I was also writing const char const * buffer, which was incorrect. const is left associative with the exception of the first type. Fortunately MSVC starting squawking about it with a warning.

I just ran out of time, so I'll have to commit this later. Thanks for the help and sorry for being so dense.

Thank you for the compliment on the quality of the code being checked in.

Reply:
Fun fact, the following three lines are all valid code:
```
const char * var;
char const * var;
char * const var;
```

But what do they mean?

Turns out the first two are equivalent.

Reply:
Before yesterday I would have said the second option meant the third option and the third option wasn't valid.